aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -1,11 +1,33 @@
 #include <ATen/native/ScatterGatherShapeChecks.h>
+#include <ATen/native/TensorAdvancedIndexing.h>
 #include <ATen/native/ReduceOpsUtils.h>
 #include <ATen/native/TensorIterator.h>
 #include <ATen/Parallel.h>
+#include <unordered_map>
 
 namespace at { namespace native {
 
 namespace {
+  
+static inline int64_t ensure_nonempty_dim(int64_t dim) {
+  return std::max<int64_t>(dim, 1);
+}
+
+static inline int64_t ensure_nonempty_size(const Tensor& t, int64_t dim) {
+  return t.dim() == 0 ? 1 : t.size(dim);
+}
+
+static inline int64_t ensure_nonempty_stride(const Tensor& t, int64_t dim) {
+  return t.dim() == 0 ? 1 : t.stride(dim);
+}
+
+using IdxVec = std::vector<int64_t>;
+static inline IdxVec ensure_nonempty_vec(IdxVec vec) {
+  if (vec.size() == 0) {
+    vec.push_back(1);
+  }
+  return vec;
+}
```

Comment:
These functions were moved to another header. Please, rebase.

Reply:
OK. Could you review the code structure? I'll implement the scalar function if its OK to proceed.

Reply:
I do not know. You might have seen I've created a PR to simplify the `cpu_scatter_gather_base_kernel`. It is in conflict with your work, unfortunately... But there is no guarantee it is going to be merged however...